README.md
# [iasc-monads][]

Thanks for starting a project with Haskeleton! If you haven't heard of it
before, I suggest reading the introductory blog post. You can find it here:
<http://taylor.fausak.me/2014/03/04/haskeleton-a-haskell-project-skeleton/>.

Before you get started, there are a few things that this template couldn't
provide for you. You should:

-   Add a synopsis to `package.yaml`. It should be a short (one sentence)
    explanation of your project.

-   Add a description to `package.yaml`. This can be whatever you want it to
    be.

-   Add a category to `package.yaml`. A list of categories is available on
    Hackage at <http://hackage.haskell.org/packages>.

-   Rename `library/Example.hs` to whatever you want your top-level module to
    be called. Typically this is the same as your package name but in
    `CamelCase` instead of `kebab-case`.

    -   Don't forget to rename the reference to it in
        `executable/Main.hs`!

-   If you are on an older version of Stack (<1.0.4), delete `package.yaml` and
    remove `/*.cabal` from your `.gitignore`.

Once you've done that, start working on your project with the Stack commands
you know and love.

``` sh
# Build the project.
stack build

# Run the test suite.
stack test

# Run the benchmarks.
stack bench

# Generate documentation.
stack haddock
```

Thanks again, and happy hacking!

[iasc-monads]: https://github.com/githubuser/iasc-monads


CHANGELOG.md
# Change log

iasc-monads uses [Semantic Versioning][].
The change log is available through the [releases on GitHub][].

[Semantic Versioning]: http://semver.org/spec/v2.0.0.html
[releases on GitHub]: https://github.com/githubuser/iasc-monads/releases


library/Example.hs
{-# LANGUAGE OverloadedStrings #-}
module Example where
import qualified Data.Aeson as JSON
import qualified Data.Aeson.Types as JSON
import Data.Text (Text, unpack)
import Text.Read
import Test.QuickCheck

getField :: (JSON.FromJSON a) => JSON.Object -> Text -> JSON.Parser a
getField = (JSON..:)

data Personaje = Personaje { nombre :: String, inventario :: [Item], rol :: Rol } deriving Show

data Rol = Warrior | Mage | Priest | DeathKnight deriving (Enum, Bounded, Show, Eq, Read)

data Item = Espada | Escudo | GrogXD deriving (Enum, Bounded, Show, Eq, Read)

data Validado a = Exito a | Error String deriving Show

count condicion = length . filter condicion

validarInventario :: [Item] -> Validado [Item]
validarInventario unInventario | count (== GrogXD) unInventario > 1 = Error "Nadie puede llevar mas de un GrogXD"
                             | otherwise = Exito unInventario

validarNombre :: String -> Validado String
validarNombre unNombre | length unNombre > 20 = Error "Nombre muy largo"
                       | otherwise = Exito unNombre

validarPersonaje :: Personaje -> Validado Personaje
validarPersonaje unPersonaje | fuerza unPersonaje <= 100 = Exito unPersonaje
                             | otherwise = Error "El personaje esta roto"

fuerza :: Personaje -> Int
fuerza unPersonaje = length (inventario unPersonaje) * indicePorRol (rol unPersonaje)
    where indicePorRol unRol = case unRol of
            Warrior -> 30
            Mage -> 30
            Priest -> 10
            DeathKnight -> 50

inicialesDePersonaje :: Personaje -> [String]
inicialesDePersonaje unPersonaje = map (take 1) (words (nombre unPersonaje))

lichKing :: Validado Personaje
lichKing = construirPersonajeValidado (validarNombre "Arthas Menethil") (validarInventario [Espada, Escudo]) (pure DeathKnight) validarPersonaje

inicialesDePersonajeValidado :: Validado Personaje -> Validado [String]
inicialesDePersonajeValidado unPersonaje = inicialesDePersonaje <$> unPersonaje

instance Functor Validado where
    fmap funcion valor = case valor of
        Exito valorValidado -> Exito (funcion valorValidado)
        Error mensajeDeError -> Error mensajeDeError

instance Applicative Validado where
    Exito funcion <*> Exito valor = Exito (funcion valor)
    Error mensajeDeError <*> _ = Error mensajeDeError
    _ <*> Error mensajeDeError = Error mensajeDeError
    pure valor = Exito valor

instance Monad Validado where
    Exito valor >>= funcion = funcion valor
    Error mensajeDeError >>= _ = Error mensajeDeError

construirPersonajeValidado :: Validado String -> Validado [Item] -> Validado Rol -> (Personaje -> Validado Personaje) -> Validado Personaje
construirPersonajeValidado nombreValidado inventarioValidado rolValidado validacionSobrePersonaje =
        (Personaje <$> nombreValidado <*> inventarioValidado <*> rolValidado) >>= validacionSobrePersonaje

posiblesPersonajes nombres inventarios roles transformaciones =
    (Personaje <$> nombres <*> inventarios <*> roles) >>= transformaciones

construirPersonaje nombre inventario rol transformacion = do
    unNombre <- nombre
    unInventario <- inventario
    unRol <- rol
    transformacion (Personaje unNombre unInventario unRol)

instance JSON.FromJSON Personaje where
    parseJSON = JSON.withObject "Personaje" $ \jsonObject ->
        construirPersonaje
            (getField jsonObject "nombre")
            (getField jsonObject "inventario")
            (getField jsonObject "rol")
            pure

instance JSON.FromJSON Item where
    parseJSON = parseReadableFromText

instance JSON.FromJSON Rol where
    parseJSON = parseReadableFromText

parseReadableFromText :: (Read a) => JSON.Value -> JSON.Parser a
parseReadableFromText = JSON.withText "Rol" $ maybe (fail "Nope") pure . readMaybe . unpack

instance Arbitrary Rol where
    arbitrary = arbitraryBoundedEnum

instance Arbitrary Item where
    arbitrary = arbitraryBoundedEnum

instance Arbitrary Personaje where
    arbitrary = construirPersonaje
                    (elements ["Arthas", "Illidan", "Thrall"])
                    arbitrary
                    arbitrary
                    pure

library/MonadExample.hs
-- la linea de abajo es para usar el typeclass Num
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module MonadExample where

-- Agregamos un validado para validar el exito o el error de una validacion
data Validado a = Exito a | Error String deriving (Show, Eq)


-- Con newtype definimos un nuevo tipo, el deriving lo dejamos mas adelante
newtype Oro = Oro Int
  deriving (Show, Eq, Num, Ord)

-- Ahora existen los items
data Item = Espada | Escudo | ScrollRojo deriving (Show, Eq)

-- Ahora estan los roles
data Rol = Warrior | Knight | Mage | Priest | DeathKnight deriving (Show, Eq)

-- ### Validaciones sobre items
count condicion = length . filter condicion

validarInventario :: [Item] -> Validado [Item]
validarInventario unInventario | count (== ScrollRojo) unInventario > 1 = Error "Nadie puede llevar mas de un scroll rojo"
                             | otherwise = Exito unInventario

-- Despues esta el personake, que debe tener un nombre para empezar
data Personaje = Personaje {
  salud :: Int,
  dinero :: Oro,
  inventario :: [Item],
  nombre :: String,
  rol :: Rol
} deriving (Show, Eq)

validarPersonaje :: Personaje -> Validado Personaje
validarPersonaje unPersonaje | fuerza unPersonaje <= 100 = Exito unPersonaje
                             | otherwise = Error "El personaje esta roto"

fuerza :: Personaje -> Int
fuerza unPersonaje = length (inventario unPersonaje) * indicePorRol (rol unPersonaje)
    where indicePorRol unRol = case unRol of 
            Warrior -> 30
            Mage -> 30
            Knight -> 35
            Priest -> 10
            DeathKnight -> 50

-- ### Funciones adicionales sobre Personaje
inicialesDePersonaje :: Personaje -> [String]
inicialesDePersonaje unPersonaje = map (take 1) (words (nombre unPersonaje))

obtenerDineroPersonaje :: Personaje -> Oro
obtenerDineroPersonaje unPersonaje = (dinero unPersonaje)

validarDinero :: Oro -> Validado Oro
validarDinero dinero | dinero >= 0 = Exito dinero
                     | otherwise = Error "Tiene que tener algo de oro o 0"

-- Ahora tenemos una funcion que nos va a validar que el nombre no sea demasiado corto 
validarNombre :: String -> Validado String
validarNombre unNombre | length unNombre < 4 = Error "El nombre es muy corto"
                       | length unNombre > 20 = Error "El nombre es muy largo"
                       | otherwise = Exito unNombre

-- #### Fxs sobre Validado Personaje

inicialesDePersonajeValidado :: Validado Personaje -> Validado [String]
--inicialesDePersonajeValidado unPersonaje = case unPersonaje of
--    Exito personajeValidado -> Exito (inicialesDePersonaje personajeValidado)
--    Error mensajeDeError -> Error mensajeDeError
-- inicialesDePersonajeValidado unPersonaje = fmap inicialesDePersonaje unPersonaje
inicialesDePersonajeValidado unPersonaje = inicialesDePersonaje <$> unPersonaje

-- Ahora queremos obtener el dinero de un personaje validado
dineroPersonajeValidado :: Validado Personaje -> Validado Oro
--dineroPersonajeValidado unPersonaje = case unPersonaje of
--    Exito personajeValidado -> Exito (obtenerDineroPersonaje personajeValidado)
--    Error mensajeDeError -> Error mensajeDeError
--dineroPersonajeValidado unPersonaje = fmap obtenerDineroPersonaje unPersonaje
dineroPersonajeValidado unPersonaje = obtenerDineroPersonaje <$> unPersonaje

instance Functor Validado where
    fmap funcion valor = case valor of
        Exito valorValidado -> Exito (funcion valorValidado)
        Error mensajeDeError -> Error mensajeDeError 

instance Applicative Validado where
    Exito funcion <*> Exito valor = Exito (funcion valor)
    Error mensajeDeError <*> _ = Error mensajeDeError
    _ <*> Error mensajeDeError = Error mensajeDeError
    pure valor = Exito valor

instance Monad Validado where
    Exito valor >>= funcion = funcion valor
    Error mensajeDeError >>= _ = Error mensajeDeError

-- La idea seria de poder crear un personaje validado...                      
construirPersonajeValidado :: Validado String -> Validado Oro -> Validado [Item] -> Validado Rol -> (Personaje -> Validado Personaje) -> Validado Personaje
construirPersonajeValidado nombreValidado plataValidada inventarioValidado rolValidado validacionSobrePersonaje
  = ((Personaje 100) <$> plataValidada <*> inventarioValidado <*> nombreValidado <*> rolValidado) >>= validacionSobrePersonaje


--- Lo mismo que construirPersonajeValidado pero con do notation
construirPersonaje nombre plata inventario rol transformacion = do
    unNombre <- nombre
    unaPlata <- plata
    unInventario <- inventario
    unRol <- rol
    transformacion (Personaje 100 unaPlata unInventario unNombre unRol)

--- Tambien se puede extraer a listas...
posiblesPersonajes nombres inventarios roles transformaciones =
    (Personaje <$> nombres <*> inventarios <*> roles) >>= transformaciones 

test-suite/Main.hs
-- Tasty makes it easy to test your code. It is a test framework that can
-- combine many different types of tests into one suite. See its website for
-- help: <http://documentup.com/feuerbach/tasty>.
import qualified Test.Tasty
-- Hspec is one of the providers for Tasty. It provides a nice syntax for
-- writing tests. Its website has more info: <https://hspec.github.io>.
import Test.Tasty.Hspec
import MonadExample

main :: IO ()
main = do
    test <- testSpec "iasc-monads" spec
    Test.Tasty.defaultMain test

spec :: Spec
spec = parallel $ do
    describe "Personaje data type" $ do
        it "Deberia poder crear un Personaje" $ do
            let nombrePersonaje = "Arthas Menethil"
            let personaje = Personaje 0 100 [] nombrePersonaje Mage
            nombre personaje `shouldBe` nombrePersonaje

        it "Dos personajes son iguales si tienen mismo nombre, dinero y salud" $ do
            let nombrePersonaje = "Arthas Menethil"
            let personaje = Personaje 0 100 [] nombrePersonaje Mage
            let otroPersonaje = Personaje 0 100 [] nombrePersonaje Mage
            personaje `shouldBe` otroPersonaje

        it "Dos personajes son distintos si tienen distintos atributos" $ do
            let nombrePersonaje = "Arthas Menethil"
            let personaje = Personaje 0 100 [] nombrePersonaje Mage
            let otroPersonaje = Personaje 200 100 [] nombrePersonaje Mage
            personaje `shouldNotBe` otroPersonaje

    describe "validarNombre" $ do
        it "Deberia fallar cuando el nombre es muy corto..." $ do
            let nombrePersonaje = "inv"
            validarNombre nombrePersonaje == Error "El nombre es muy corto"

        it "Deberia fallar cuando el nombre es demasiado largo..." $ do
            let nombrePersonaje = "esteNoEsUnNOmbreValidoAlSerDemasiadoLargo"
            validarNombre nombrePersonaje == Error "El nombre es muy largo" 

        it "Deberia ser exitosa la validacion" $ do
            let nombrePersonaje = "Arthas Menethil"
            validarNombre nombrePersonaje == Exito nombrePersonaje
    
    describe "construirPersonajeValidado" $ do
        it "Deberia fallar cuando el nombre es muy corto..." $ do
            let nombrePersonaje = validarNombre "inv"
            let plata = validarDinero 100
            (construirPersonajeValidado nombrePersonaje plata (Exito []) (pure Knight) validarPersonaje) == Error "El nombre es muy corto"

        it "Deberia fallar cuando el nombre es demasiado largo..." $ do
            let nombrePersonaje = validarNombre "esteNoEsUnNOmbreValidoAlSerDemasiadoLargo"
            let plata = validarDinero 100
            (construirPersonajeValidado nombrePersonaje plata (Exito []) (pure Knight) validarPersonaje) == Error "El nombre es muy largo"      

        it "Deberia fallar cuando el inventario es invalido" $ do
            let nombrePersonaje = validarNombre "JuanWizard"
            let plata = validarDinero 250
            let items = validarInventario [ScrollRojo, Espada, ScrollRojo]
            (construirPersonajeValidado nombrePersonaje plata items (pure Mage) validarPersonaje) == Error "Nadie puede llevar mas de un scroll rojo"

        it "Deberia crearme el personaje cuando el nombre tiene la longitud correcta" $ do
            let nombrePersonaje = validarNombre "NicoKnight"
            let items = validarInventario [Escudo]
            let plata = validarDinero 250
            (construirPersonajeValidado nombrePersonaje plata items (pure Knight) validarPersonaje) == Exito (Personaje 100 250 [Escudo] "NicoKnight" Knight)

        it "Deberia fallar cuando el personaje esta roto" $ do
            let nombrePersonaje = validarNombre "NicoKnight"
            let items = validarInventario [ScrollRojo, Escudo, Espada]
            let plata = validarDinero 250
            (construirPersonajeValidado nombrePersonaje plata items (pure Knight) validarPersonaje) == Error "El personaje esta roto"

    describe "F(x)s adicionales" $ do
        it "deberia devolverme el plata seteado al personaje" $ do
            let nombrePersonaje = "Arthas Menethil"
            let personaje = Personaje 100 0 [] nombrePersonaje DeathKnight
            obtenerDineroPersonaje personaje `shouldBe` 0

        it "deberia devolverme las iniciales correctas del personaje" $ do
            let nombrePersonaje = "Arthas Menethil"
            let personaje = Personaje 100 0 [] nombrePersonaje DeathKnight
            (inicialesDePersonaje personaje) `shouldBe` ["A","M"]

    describe "inicialesDePersonajeValidado" $ do
        it "Deberia obtener las iniciales de un personaje validado" $ do
            let personajeValidado = construirPersonajeValidado (validarNombre "Arthas Menethil") (Exito 1000) (Exito []) (pure DeathKnight) validarPersonaje
            inicialesDePersonajeValidado personajeValidado == Exito (["A","M"])

        it "Deberia fallar al ser el nombre del personaje muy corto" $ do
            let personajeValidado = construirPersonajeValidado (validarNombre "inv") (Exito 1000) (Exito []) (pure DeathKnight) validarPersonaje
            inicialesDePersonajeValidado personajeValidado == Error "El nombre es muy corto"

    describe "dineroPersonajeValidado" $ do
        it "Deberia obtener el saldo de un personaje validado" $ do
            let personajeValidado = construirPersonajeValidado (validarNombre "Arthas Menethil") (Exito 1000) (Exito []) (pure DeathKnight) validarPersonaje
            dineroPersonajeValidado personajeValidado == Exito (1000)

        it "Deberia fallar al ser el nombre del personaje muy corto" $ do
            let personajeValidado = construirPersonajeValidado (validarNombre "inv") (Exito 1000) (Exito []) (pure DeathKnight) validarPersonaje
            dineroPersonajeValidado personajeValidado == Error "El nombre es muy corto"

    describe "validarInventario" $ do
        it "cuando hay mas de un scroll rojo en la lista de items" $ do
            let items = [ScrollRojo, Espada, ScrollRojo]
            validarInventario items == Error "Nadie puede llevar mas de un scroll rojo"

        it "cuando se puede validar sin problemas la lista de items" $ do
            let items = [ScrollRojo, Espada, Escudo]
            validarInventario items == Exito items

    describe "validarDinero" $ do
        it "deberia fallar al ser un numero negativo" $ do
            validarDinero (-1) `shouldBe` Error "Tiene que tener algo de oro o 0"

        it "deberia ser un exito al ser un numero positivo" $ do
            validarDinero (100) `shouldBe` Exito (100)

    describe "fuerza" $ do
        it "Deberia dar 0" $ do
            let personaje = Personaje 100 0 [] "caballero3842" DeathKnight
            fuerza personaje `shouldBe` 0

        it "Deberia dar 50 para el DeathKnight con un item" $ do
            let personaje = Personaje 100 0 [ScrollRojo] "caballero3842" DeathKnight
            fuerza personaje `shouldBe` 50

        it "Deberia dar 150 para el DeathKnight con tres items" $ do
            let personaje = Personaje 100 10 [ScrollRojo, Escudo, Espada] "caballero3842" DeathKnight
            fuerza personaje `shouldBe` 150

        it "Deberia dar 30 para el Mage con un item" $ do
            let personaje = Personaje 100 0 [ScrollRojo] "mage132" Mage
            fuerza personaje `shouldBe` 30
    
    describe "validarPersonaje" $ do
        it "deberia fallar al ser la fuerza mayor a 100" $ do
            let personaje = Personaje 100 10 [ScrollRojo, Escudo, Espada] "caballero3842" DeathKnight
            validarPersonaje personaje `shouldBe` Error "El personaje esta roto"

        it "deberia validarse al ser la fuerza menor a 100" $ do
            let personaje = Personaje 100 0 [ScrollRojo] "mage132" Mage
            validarPersonaje personaje `shouldBe` Exito personaje
    
    describe "construirPersonaje" $ do
        it "Deberia fallar cuando el nombre es demasiado largo..." $ do
            let nombrePersonaje = validarNombre "esteNoEsUnNOmbreValidoAlSerDemasiadoLargo"
            let plata = validarDinero 100
            (construirPersonaje nombrePersonaje plata (Exito []) (pure Knight) validarPersonaje) == Error "El nombre es muy largo"      

        it "Deberia crearme el personaje cuando el nombre tiene la longitud correcta" $ do
            let nombrePersonaje = validarNombre "NicoKnight"
            let items = validarInventario [Escudo]
            let plata = validarDinero 250
            (construirPersonaje nombrePersonaje plata items (pure Knight) validarPersonaje) == Exito (Personaje 100 250 [Escudo] "NicoKnight" Knight)

        it "Deberia fallar cuando el personaje esta roto" $ do
            let nombrePersonaje = validarNombre "NicoKnight"
            let items = validarInventario [ScrollRojo, Escudo, Espada]
            let plata = validarDinero 250
            (construirPersonaje nombrePersonaje plata items (pure Knight) validarPersonaje) == Error "El personaje esta roto"
